types/src/v0/impls/reward.rs

```
@@ -0,0 +1,425 @@
+                );
+                Ok(proof
+                    .elem()
+                    .context("presence proof is missing account balance")?
+                    .0)
+            },
+            RewardMerkleProof::Absence(proof) => {
+                let tree = RewardMerkleTree::from_commitment(comm);
+                ensure!(
+                    tree.non_membership_verify(RewardAccount(self.account), proof)?,
+                    "invalid proof"
+                );
+                Ok(0.into())
+            },
+        }
+    }
+
+    pub fn remember(&self, tree: &mut RewardMerkleTree) -> anyhow::Result<()> {
+        match &self.proof {
+            RewardMerkleProof::Presence(proof) => {
+                tree.remember(
+                    RewardAccount(self.account),
+                    proof
+                        .elem()
+                        .context("presence proof is missing account balance")?,
+                    proof,
+                )?;
+                Ok(())
+            },
+            RewardMerkleProof::Absence(proof) => {
+                tree.non_membership_remember(RewardAccount(self.account), proof)?;
+                Ok(())
+            },
+        }
+    }
+}
+
+impl From<(RewardAccountProof, U256)> for RewardAccountQueryData {
+    fn from((proof, balance): (RewardAccountProof, U256)) -> Self {
+        Self { balance, proof }
+    }
+}
+
+pub fn apply_rewards(
+    mut reward_state: RewardMerkleTree,
+    validator: Validator<BLSPubKey>,
+) -> anyhow::Result<RewardMerkleTree> {
+    let mut update_balance = |account: &RewardAccount, amount: RewardAmount| {
+        reward_state = reward_state.persistent_update_with(account, |balance| {
+            let balance = balance.copied();
+            match balance.unwrap_or_default().0.checked_add(amount.0) {
+                Some(updated) => Some(updated.into()),
+                None => {
+                    tracing::warn!("overflowed reward balance for account {}", account);
```

This seems like a really unexpected error. Should we really continue? It means we cannot actually calculate the correct rewards so it's a breakdown of the protocol. Shouldn't we return an error, panic or whatever.

---

https://github.com/EspressoSystems/espresso-network/pull/2772/commits/0db1340254018a76e0ec25ad91351e9576a4969c